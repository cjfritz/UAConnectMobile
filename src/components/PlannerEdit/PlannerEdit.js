import React, { Component } from 'react';
import {
  Container, Content, Button, H3,
} from 'native-base';
import { Alert, View } from 'react-native';
import { forEach } from 'lodash';
import { connect } from 'react-redux';
import PlannerFormComponent from '../PlannerForm/PlannerForm';
import {
  plannerUpdate, plannerSave, plannerDelete, plannerValidUpdate,
} from '../../actions/PlannerActions';
import styles from './PlannerEdit.style';

export class PlannerEdit extends Component {
  static navigationOptions = {
    title: 'Edit Course',
    headerStyle: {
      backgroundColor: 'darkred',
    },
    headerTitleStyle: {
      color: 'white',
    },
  };

  componentWillMount() {
    const { navigation, coursePlannerUpdate, updateValidFields } = this.props;
    const courseItem = navigation.getParam('courseItem', {});
    forEach(courseItem, (value, prop) => {
      coursePlannerUpdate({ prop, value });
      updateValidFields({ prop, value });
    });
  }

  onSave = () => {
    Alert.alert(
      'Save Changes?',
      undefined,
      [
        { text: 'Cancel' },
        { text: 'Confirm', onPress: () => this.onConfirmSave() },
      ]
    );
  };

  onConfirmSave = () => {
    const {
      course,
      description,
      term,
      grade,
      units,
      coursePlannerSave,
      navigation,
    } = this.props;
    const courseItem = navigation.getParam('courseItem', {});
    coursePlannerSave({
      uid: courseItem.uid,
      course,
      description,
      term,
      grade,
      units,
    });
  };

  onDelete = () => {
    Alert.alert(
      'Delete Course?',
      undefined,
      [
        { text: 'Cancel' },
        { text: 'Confirm', onPress: () => this.onConfirmDelete() },
      ]
    );
  };

  onConfirmDelete = () => {
    const { coursePlannerDelete, navigation } = this.props;
    coursePlannerDelete(navigation.getParam('courseItem', {}).uid);
  };

  isValid = () => {
    const { validFields } = this.props;
    const {
      validCourse,
      validDescription,
      validTerm,
      validGrade,
      validUnits,
    } = validFields;

    if (validCourse && validDescription && validTerm && validGrade && validUnits) {
      return true;
    } else {
      return false;
    }
  };

  render() {
    return (
      <Container>
        <Content>
          <PlannerFormComponent />
          <View style={ styles.buttonView }>
            <Button style={ styles.deleteButton } onPress={ () => this.onDelete() }>
              <H3 style={ styles.deleteText }>
                Delete
              </H3>
            </Button>
            <Button disabled={ !this.isValid() } style={ styles.saveButton } onPress={ () => this.onSave() }>
              <H3 style={ styles.saveText }>
                Save
              </H3>
            </Button>
          </View>
        </Content>
      </Container>
    );
  }
}

const mapDispatchToProps = dispatch => (
  ({
    coursePlannerUpdate: ({ prop, value }) => dispatch(plannerUpdate({ prop, value })),
    coursePlannerSave: ({
      uid,
      course,
      description,
      term,
      grade,
      units,
    }) => dispatch(plannerSave({
      uid,
      course,
      description,
      term,
      grade,
      units,
    })),
    coursePlannerDelete: uid => dispatch(plannerDelete(uid)),
    updateValidFields: ({ prop, value }) => dispatch(plannerValidUpdate({ prop, value })),
  })
);

const mapStateToProps = state => {
  const {
    course,
    description,
    term,
    grade,
    units,
    validFields,
  } = state.plannerForm;
  return {
    course,
    description,
    term,
    grade,
    units,
    validFields,
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PlannerEdit);
