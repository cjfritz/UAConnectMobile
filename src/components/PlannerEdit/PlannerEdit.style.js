import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  buttonView: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-around',
    marginTop: '5%',
  },
  deleteButton: {
    alignItems: 'center',
    backgroundColor: 'darkred',
    justifyContent: 'center',
    width: '40%',
  },
  deleteText: {
    color: 'white',
  },
  saveButton: {
    alignItems: 'center',
    backgroundColor: 'white',
    justifyContent: 'center',
    width: '40%',
  },
  saveText: {
    color: 'darkred',
  },
});
