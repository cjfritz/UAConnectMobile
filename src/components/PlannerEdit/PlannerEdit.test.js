import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { PlannerEdit } from './PlannerEdit';

const props = {
  course: '',
  description: '',
  term: '',
  grade: '',
  units: '',
  coursePlannerUpdate: jest.fn(),
  coursePlannerSave: jest.fn(),
  coursePlannerDelete: jest.fn(),
  navigation: {
    getParam: jest.fn(),
  },
  validFields: {
    validCourse: false,
    validDescription: false,
    validTerm: false,
    validGrade: false,
    validUnits: false,
  },
};

describe('<PlannerEdit />', () => {
  it('can render with empty, invalid props', () => {
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerEdit { ...props } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with invalid text for grade and units props', () => {
    const newProps = {
      grade: 'badInput',
      units: 'badInput',
      validFields: {
        validCourse: true,
        validDescription: true,
        validTerm: true,
      },
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerEdit { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with invalid values greater than max range for grade and units props', () => {
    const newProps = {
      grade: 4.1,
      units: 5,
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerEdit { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with invalid values less than min range for grade and units props', () => {
    const newProps = {
      grade: -0.5,
      units: -1,
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerEdit { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });

  it('can render with valid values for props', () => {
    const newProps = {
      course: 'testCourse',
      description: 'testDescription',
      term: 'testTerm',
      grade: '',
      units: '',
      validFields: {
        validCourse: true,
        validDescription: true,
        validTerm: true,
        validGrade: true,
        validUnits: true,
      },
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<PlannerEdit { ...props } { ...newProps } />);
    expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
  });
});
