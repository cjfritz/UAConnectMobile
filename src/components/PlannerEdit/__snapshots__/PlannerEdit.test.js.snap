// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlannerEdit /> can render with empty, invalid props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginTop": "5%",
        }
      }
    >
      <Styled(Button)
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Delete
        </Styled(H3)>
      </Styled(Button)>
      <Styled(Button)
        disabled={true}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "darkred",
            }
          }
        >
          Save
        </Styled(H3)>
      </Styled(Button)>
    </Component>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerEdit /> can render with invalid text for grade and units props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginTop": "5%",
        }
      }
    >
      <Styled(Button)
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Delete
        </Styled(H3)>
      </Styled(Button)>
      <Styled(Button)
        disabled={true}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "darkred",
            }
          }
        >
          Save
        </Styled(H3)>
      </Styled(Button)>
    </Component>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerEdit /> can render with invalid values greater than max range for grade and units props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginTop": "5%",
        }
      }
    >
      <Styled(Button)
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Delete
        </Styled(H3)>
      </Styled(Button)>
      <Styled(Button)
        disabled={true}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "darkred",
            }
          }
        >
          Save
        </Styled(H3)>
      </Styled(Button)>
    </Component>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerEdit /> can render with invalid values less than min range for grade and units props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginTop": "5%",
        }
      }
    >
      <Styled(Button)
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Delete
        </Styled(H3)>
      </Styled(Button)>
      <Styled(Button)
        disabled={true}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "darkred",
            }
          }
        >
          Save
        </Styled(H3)>
      </Styled(Button)>
    </Component>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<PlannerEdit /> can render with valid values for props 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Connect(PlannerForm) />
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginTop": "5%",
        }
      }
    >
      <Styled(Button)
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "white",
            }
          }
        >
          Delete
        </Styled(H3)>
      </Styled(Button)>
      <Styled(Button)
        disabled={false}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "justifyContent": "center",
            "width": "40%",
          }
        }
      >
        <Styled(H3)
          style={
            Object {
              "color": "darkred",
            }
          }
        >
          Save
        </Styled(H3)>
      </Styled(Button)>
    </Component>
  </Styled(Content)>
</Styled(Container)>
`;
